### ISSUE TITLE
Are file uploads changed in request to cause content hash mismatch?

### ISSUE THREAD

**Issue by hmistry** [httprb/http#423]
I'm using HTTPrb to upload files to AWS S3 and I'm getting a content hash mismatch between what I submit and what AWS computes for the file. File is a simple "Lorem ipsum..." ascii test file.

If I send the same file using curl with the same `ClientComputedContentSHA256`, it works!

Does sending the file as a form change it or produce a different data body attached to the request than curl's -T? 

Is there another way to send files using HTTPrb that doesn't modify the request body similar to curl's -T? Or if not, I can do a hash on request content after HTTPrb has attached the file and  use that for the content header but not sure how I'd do that?

```
# Working Curl command
curl -v -T "download/test-000001.txt"  http://mybucket.s3.amazonaws.com/download/test-000001.txt -H "authorization: AWS4-HMAC-SHA256 Credential=****, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=****" -H "x-amz-content-sha256: aa317e490****" -H "x-amz-date: 20170726T070855Z" 
``` 

```ruby
def hexdigest(value)
  OpenSSL::Digest::SHA256.file(value).hexdigest
end

# This does not work!
# Hexdigest of file is computed and submitted as x-amz-content-sha256 header
req = HTTP.headers(aws_auth_headers(datetime_now, "PUT", S3_URL, "", hexdigest(file)))
req.put("http://#{S3_URL}/#{file.path}", :form => { :media => HTTP::FormData::File.new(file.path) })
```
```xml
<Error>
  <Code>XAmzContentSHA256Mismatch</Code>
  <Message>The provided 'x-amz-content-sha256' header does not match what was computed.</Message>
  <ClientComputedContentSHA256>aa317e490****</ClientComputedContentSHA256>
  <S3ComputedContentSHA256>ce8e99c01****</S3ComputedContentSHA256>
</Error>
```

**Comment by tarcieri** [comment]
This would be a pretty severe problem if true... can you perhaps put together a self-contained reproduction that doesn't involve Amazon?

**Comment by ixti** [comment]
`curl -T` sends given file AS IS. It's not wrapping it up into a form. What you want is:

``` ruby
req = HTTP.headers(aws_auth_headers(datetime_now, "PUT", S3_URL, "", hexdigest(file)))
req.put("http://#{S3_URL}/#{file.path}", :body => file)
```

**Comment by ixti** [comment]
Here's what remote server will receive if you ill run `curl -T foo http://localhost:8080`:

```
PUT /foo HTTP/1.1
Host: localhost:8080
User-Agent: curl/7.53.0
Accept: */*
Content-Length: 15
Expect: 100-continue

Lorem ipsum...
```

When you pass `:form` to http ruby client:

``` ruby
HTTP.put("http://localhost:8080", :form => {
  :media => HTTP::FormData::File.new("foo")
})
```

It will send a multipart form-data body:

```
PUT / HTTP/1.1
Connection: close
Content-Type: multipart/form-data; boundary=---------------------ffb5de5c1a58210cea36b6dfb36214b1d28b7c3277
Content-Length: 255
Host: localhost:8080
User-Agent: http.rb/2.2.2

-----------------------ffb5de5c1a58210cea36b6dfb36214b1d28b7c3277
Content-Disposition: form-data; name="media"; filename="foo"
Content-Type: application/octet-stream

Lorem ipsum...

-----------------------ffb5de5c1a58210cea36b6dfb36214b1d28b7c3277--
```

**Comment by hmistry** [comment]
@ixti Yes I figured out later that it was wrapping it in a form vs just attaching it.

Is there a complete list of different options to pass in a request? As a first time user of HTTPrb, the only way I learned about how to attach file using form was because I came across it in the repo's issue list! If you have a listing of all the different options some where, I can spare time to turn it into a wiki if you'd like.

 @tarcieri It was wrapping the file in a form as shown above.
